code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,461 @@
+	return (locate(typepath) in traits)
+
+/obj/item/xenoartifact/proc/generate_icon(var/icn, var/icnst = "", colour) //Add extra icon overlays. Ghetto GAGS
+	icon_overlay = mutable_appearance(icn, icnst)
+	icon_overlay.layer = FLOAT_LAYER //Not doing this fucks the object icons when you're holding it
+	icon_overlay.appearance_flags = RESET_ALPHA// Not doing this fucks the alpha?
+	icon_overlay.alpha = alpha
+	if(colour)
+		icon_overlay.color = colour
+	add_overlay(icon_overlay)
+
+/obj/item/xenoartifact/proc/process_target(atom/target) //Used for hand-holding secret technique. Pulling entities swaps them for you in the target list.
+	. = target
+	if(isliving(target?.loc))
+		. = target?.loc
+	var/mob/living/M = istype(target, /mob/living) ? target : null //Have to type convert to access pulling
+	if(M && M?.pulling)
+		. = M?.pulling
+	RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/on_target_del, target)
+	return
+
+/obj/item/xenoartifact/proc/on_target_del(atom/target) //Hard del handle
+	UnregisterSignal(target, COMSIG_PARENT_QDELETING)
+	true_target -= target
+	target = null
+
+/obj/item/xenoartifact/proc/create_beam(atom/target) //Helps show how the artifact is working. Hint stuff.
+	if(isliving(loc) || !get_turf(target))
+		return
+	var/datum/beam/xenoa_beam/B = new(src.loc, target, time=1.5 SECONDS, beam_icon='icons/obj/xenoarchaeology/xenoartifact.dmi', beam_icon_state="xenoa_beam", btype=/obj/effect/ebeam/xenoa_ebeam)
+	B.set_color(material)
+	INVOKE_ASYNC(B, /datum/beam/xenoa_beam.proc/Start)
+
+/obj/item/xenoartifact/proc/default_activate(chr, mob/user, atom/target) //Default template used to interface with activator signals. Handle
+	if(!COOLDOWN_FINISHED(src, xenoa_cooldown))
+		return FALSE
+	charge = chr
+	true_target = !(locate(target) in true_target) ? true_target+process_target(target) : true_target
+	check_charge(user)
+	return TRUE
+
+/obj/item/xenoartifact/proc/set_frequency(new_frequency) //Signaler traits
+	SSradio.remove_object(src, frequency)
+	frequency = new_frequency
+	radio_connection = SSradio.add_object(src, frequency, RADIO_SIGNALER)
+
+/obj/item/xenoartifact/proc/send_signal(var/datum/signal/signal)
+	if(!radio_connection||!signal)
+		return
+	radio_connection.post_signal(src, signal)
+
+/obj/item/xenoartifact/receive_signal(datum/signal/signal)
+	if(!signal || signal.data["code"] != code)
+		return
+	SEND_SIGNAL(src, XENOA_SIGNAL, null, get_proximity(max_range), get_proximity(max_range)) //I don't think this sends a signal
+
+/obj/item/xenoartifact/on_block(mob/living/carbon/human/owner, atom/movable/hitby)
+	. = ..()
+	if(!(COOLDOWN_FINISHED(src, xenoa_cooldown)) || !get_trait(/datum/xenoartifact_trait/minor/blocking))
+		return
+	SEND_SIGNAL(src, COMSIG_PARENT_ATTACKBY, src, owner, hitby) //I don't think this sends a signal
+
+/obj/item/xenoartifact/process(delta_time)
+	switch(process_type)
+		if(IS_LIT) //Bruning
+			true_target = list(get_proximity(min(max_range, 5)))
```

Comment:
this method could be changed to be `get_viable_targets` or `get_targets_in_proximity`